src/context/FoxEthProvider/FoxEthProvider.tsx

```diff
@@ -0,0 +1,297 @@
+  const [foxFarmingOpportunities, setFoxFarmingOpportunities] = useState<
+    FoxFarmingEarnOpportunityType[]
+  >([v4FarmingOpportunity])
+  const { isLpAprLoaded, lpApr } = useLpApr()
+  const ethMarketData = useAppSelector(state => selectMarketDataById(state, ethAssetId))
+  const foxMarketData = useAppSelector(state => selectMarketDataById(state, foxAssetId))
+  const foxAssetPrecision = useAppSelector(state => selectAssetById(state, foxAssetId)).precision
+  const lpAssetPrecision = useAppSelector(state =>
+    selectAssetById(state, foxEthLpAssetId),
+  ).precision
+
+  useEffect(() => {
+    if (wallet && adapter) {
+      ;(async () => {
+        const address = await adapter.getAddress({ wallet })
+        setConnectedWalletEthAddress(address)
+      })()
+    }
+  }, [adapter, wallet])
+
+  const reloadFarmingOpportunities = useCallback(async () => {
+    if (!connectedWalletEthAddress) return
+    try {
+      const newOpportunities = await Promise.all(
+        foxFarmingOpportunities.map(async opportunity => {
+          const data = await getOpportunityData({
+            contractAddress: opportunity.contractAddress,
+            ethAssetPrecision: ethAsset.precision,
+            ethPrice: ethMarketData.price,
+            lpAssetPrecision,
+            address: connectedWalletEthAddress,
+            foxPrice: foxMarketData.price,
+            foxAssetPrecision,
+          })
+          if (!data) return opportunity
+          const { tvl, apr, balances, expired } = data
+          return {
+            ...opportunity,
+            cryptoAmount: balances.cryptoBalance,
+            fiatAmount: balances.fiatBalance,
+            isLoaded: isLpAprLoaded,
+            unclaimedRewards: balances.unclaimedRewards,
+            expired,
+            apy: lpApr
+              ? bnOrZero(apr)
+                  .plus(lpApr ?? 0)
+                  .toString()
+              : undefined,
+            tvl,
+          }
+        }),
+      )
+      const totalOpBalances = newOpportunities.reduce(
+        (acc, cur) => acc.plus(bnOrZero(cur.fiatAmount)),
+        bnOrZero(0),
+      )
+      setFoxFarmingTotalBalance(totalOpBalances.toFixed(2))
+      setFoxFarmingOpportunities(newOpportunities)
+    } catch (error) {
+      console.error('error', error)
+    } finally {
+      setFarmingLoading(false)
+    }
+  }, [
+    connectedWalletEthAddress,
+    foxFarmingOpportunities,
+    ethAsset.precision,
+    ethMarketData.price,
+    lpAssetPrecision,
+    foxMarketData.price,
+    foxAssetPrecision,
+    isLpAprLoaded,
+    lpApr,
+  ])
+
+  const reloadLpOpportunity = useCallback(async () => {
+    try {
+      const balances = await calculateHoldings()
+      if (balances) {
+        const { lpBalance, foxBalance, ethBalance } = balances
+        setLpFoxBalance(foxBalance.toString())
+        setLpEthBalance(ethBalance.toString())
+        const totalLpBalance = bnOrZero(lpBalance).div(`1e${lpAssetPrecision}`)
```

Comment:
- `lpBalance` defaults to zero:

```suggestion
        const totalLpBalance = bn(lpBalance).div(`1e${lpAssetPrecision}`)
```

- nitpick, but this doesn't lose type-safety:

```suggestion
        const totalLpBalance = bn(lpBalance).div(bn(10).pow(lpAssetPrecision))
```